packages/flutter/test/rendering/paragraph_test.dart

```diff
@@ -36,6 +36,42 @@ void main() {
     expect(offset50.dy, greaterThan(offset5.dy));
   });
 
+  test('preferredLineHeight control test', () {
+    final RenderParagraph paragraph = RenderParagraph(
+      const TextSpan(text: _kText),
+      textDirection: TextDirection.ltr,
+    );
+
+    expect(paragraph.preferredLineHeight, greaterThan(0.0));
+  });
+
+  test('getFullHeightForCaret control test', () {
+    final RenderParagraph paragraph = RenderParagraph(
+      const TextSpan(text: _kText),
+      textDirection: TextDirection.ltr,
+    );
+    layout(paragraph);
+
+    const Rect caret = Rect.fromLTWH(0.0, 0.0, 2.0, 20.0);
+
+    final double offset5 = paragraph.getFullHeightForCaret(const TextPosition(offset: 5), caret);
+    expect(offset5, greaterThan(0.0));
+  });
+
+  test('getLineBoundary control test', () {
+    final RenderParagraph paragraph = RenderParagraph(
+      const TextSpan(text: _kText),
+      textDirection: TextDirection.ltr,
+    );
+    layout(paragraph);
+
+    final TextRange range5 = paragraph.getLineBoundary(const TextPosition(offset: 5));
+    expect(range5.textInside(_kText), equals('I polished up that handle so carefullee'));
```

Comment:
The doc on `TextPainter` says "The newline, if any, is included in the range". However this doesn't seem to be the case... or I missunderstand this statement.

I couldn't find any tests for `TextPainter.getLineBoundary` to see if this is expected.

Reply:
Oh, interesting. So the web tests failed here with:

```
07:30 +485 ~12 -1: /b/s/w/ir/k/flutter/packages/flutter/test/rendering/paragraph_test.dart: getLineBoundary control test [E]                                                                           
  Expected: 'I polished up that handle so carefullee'
    Actual: 'I polished up that handle so carefullee\n'
              ''
     Which: is different. Both strings start the same, but the actual value also has the following trailing characters: \n
```

Which sounds like implementation for the web platform is more aligned with the doc... ?

Running this test locally on macOS works fine.

I'm probably going to need help determining which behavior is the correct one and whether I should be looking at the TextPainter implementation to fix this inconsistency.

Reply:
I would agree with you, something seems not right there.  Let's see if @GaryQian knows what the behavior should be.

There are some tests in the web part of the engine for this, and they include the newline: https://github.com/flutter/engine/blob/09a5bf7acf44dbb22d68830f5a51ebe822d5e659/lib/web_ui/test/paragraph_test.dart#L926

Reply:
This may indeed be a bug, I think the newline makes sense here, especially since the docs indicate that it was meant to be included. It isn't your responsibility to fix that though, so you may want to just mark it with a comment and a `TODO: blahblah` to indicate that the result here may be due to a bug. Then, you can duplicate the test and make one of them skip web while having the other skip android+ios. We should file a bug on this.